File: tests/oracle/merge_datatype_scale_1204.cc
/* MERGE-DATATYPE-SCALE-0001: locked Ghidra 12.0.4 production-range oracle. */
#include <bits/stdc++.h>

#define private public
#include "architecture.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef private

using namespace ghidra;
using std::cerr;
using std::cout;
using std::istringstream;
using std::map;
using std::ostringstream;
using std::set;
using std::string;
using std::vector;

namespace {

class FixtureTranslate final : public Translate {
  VarnodeData dummyRegister;
public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this,this));
    insertSpace(new AddrSpace(this,this,IPTR_PROCESSOR,"other",false,8,1,1,
                              AddrSpace::hasphysical,0,0));
    insertSpace(new UniqueSpace(this,this,2,0));
    AddrSpace *ram = new AddrSpace(this,this,IPTR_PROCESSOR,"ram",false,8,1,3,
                                   AddrSpace::hasphysical,0,0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this,this,IPTR_PROCESSOR,"register",false,8,1,4,
                                   AddrSpace::hasphysical,0,0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(this,this,"stack",5,8,ram,1,true);
    insertSpace(stack);
    VarnodeData stackPointer = { reg, 0, 8 };
    addSpacebasePointer(stack,stackPointer,8,true);
    insertSpace(new AddrSpace(this,this,IPTR_PROCESSOR,"join",false,8,1,6,
                              AddrSpace::hasphysical,0,0));
    insertSpace(new IopSpace(this,this,7));
    setDefaultCodeSpace(3);
    dummyRegister = { reg, 0, 8 };
  }
  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const string &) const override { return dummyRegister; }
  string getRegisterName(AddrSpace *,uintb,int4) const override { return ""; }
  string getExactRegisterName(AddrSpace *,uintb,int4) const override { return ""; }
  void getAllRegisters(map<VarnodeData,string> &) const override {}
  void getUserOpNames(vector<string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &,const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &,const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override { return (PcodeInjectLibrary *)0; }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}
public:
  FixtureArchitecture() {
    FixtureTranslate *fixtureTranslate = new FixtureTranslate();
    translate = fixtureTranslate;
    copySpaces(fixtureTranslate);
    max_basetype_size = 16;
    types = new TypeFactory(this);
    types->setupSizes();
    types->setCoreType("xunknown1",1,TYPE_UNKNOWN,false);
    types->setCoreType("xunknown2",2,TYPE_UNKNOWN,false);
    types->setCoreType("xunknown4",4,TYPE_UNKNOWN,false);
    types->setCoreType("xunknown8",8,TYPE_UNKNOWN,false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst,types,translate);
    symboltab = new Database(this,false);
    symboltab->attachScope(new ScopeInternal(0x101,"",this),(Scope *)0);
    ProtoModel *model = new ProtoModel(this);
    istringstream stream("<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }
  void printMessage(const string &) const override {}
};

Varnode *makeWritten(Funcdata &fd,BlockBasic *block,AddrSpace *space,uintb offset,
                     uintb pc,Datatype *type)
{
  PcodeOp *op = fd.newOp(0,Address(fd.getArch()->getDefaultCodeSpace(),pc));
  fd.opSetOpcode(op,CPUI_COPY);
  Varnode *vn = fd.newVarnodeOut(4,Address(space,offset),op);
  vn->updateType(type);
  fd.opInsertEnd(op,block);
  return vn;
}

string instanceSpaces(HighVariable *high)
{
  ostringstream stream;
  for(int4 i=0;i<high->numInstances();++i) {
    if (i != 0) stream << '/';
    stream << high->getInstance(i)->getSpace()->getIndex();
  }
  return stream.str();
}

void runScale(size_t freeCount)
{
  FixtureArchitecture architecture;
  AddrSpace *unique = architecture.getSpace(2);
  AddrSpace *ram = architecture.getSpace(3);
  AddrSpace *reg = architecture.getSpace(4);
  Scope *global = architecture.symboltab->getGlobalScope();
  Funcdata fd("merge_scale","merge_scale",global,Address(ram,0x8000),
              (FunctionSymbol *)0,0x100);
  BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
  BlockBasic *block = blocks.newBlockBasic(&fd);
  fd.setBasicBlockRange(block,Address(ram,0x8000),Address(ram,0x80ff));

  Datatype *common = architecture.types->getBase(4,TYPE_UINT);
  TypeBase *identityA = new TypeBase(4,TYPE_UINT,"identity_same");
  TypeBase *identityB = new TypeBase(4,TYPE_UINT,"identity_same");
  TypeBase *tieType = new TypeBase(4,TYPE_UINT,"same_address_tie");

  vector<Varnode *> freeValues;
  for(size_t i=0;i<freeCount;++i) {
    Varnode *vn = fd.newVarnode(4,Address(reg,0x1000 + i * 8));
    vn->updateType(common);
    freeValues.push_back(vn);
  }

  Varnode *uniqueValue = makeWritten(fd,block,unique,0x30,0x8000,common);
  Varnode *ramValue = makeWritten(fd,block,ram,0x20,0x8001,common);
  Varnode *registerValue = makeWritten(fd,block,reg,0x10,0x8002,common);
  Varnode *implied = makeWritten(fd,block,reg,0x40,0x8003,common);
  implied->setImplied();
  Varnode *protoPartial = makeWritten(fd,block,reg,0x50,0x8004,common);
  protoPartial->setProtoPartial();
  Varnode *spacebase = makeWritten(fd,block,reg,0x60,0x8005,common);
  spacebase->setFlags(Varnode::spacebase);
  Varnode *typeA = makeWritten(fd,block,reg,0x70,0x8006,identityA);
  Varnode *typeB = makeWritten(fd,block,reg,0x80,0x8007,identityB);
  Varnode *tieFirst = makeWritten(fd,block,reg,0x90,0x8008,tieType);
  Varnode *tieSecond = makeWritten(fd,block,reg,0x90,0x8009,tieType);

  fd.setHighLevel();
  HighVariable *expectedSurvivor = uniqueValue->getHigh();
  HighVariable *expectedTieSurvivor = tieFirst->getHigh();
  fd.getMerge().mergeByDatatype(fd.beginLoc(),fd.endLoc());

  set<HighVariable *> freeHighs;
  bool freeSingletons = true;
  for(Varnode *vn : freeValues) {
    freeHighs.insert(vn->getHigh());
    freeSingletons &= vn->getHigh()->numInstances() == 1;
  }
  HighVariable *merged = uniqueValue->getHigh();
  bool legalMerged = merged == ramValue->getHigh() && merged == registerValue->getHigh();
  bool survivor = merged == expectedSurvivor;
  bool filtersSingleton = implied->getHigh()->numInstances() == 1 &&
      protoPartial->getHigh()->numInstances() == 1 &&
      spacebase->getHigh()->numInstances() == 1;
  bool identitySeparate = typeA->getHigh() != typeB->getHigh();
  HighVariable *tieMerged = tieFirst->getHigh();
  bool sameAddressStable = tieMerged == expectedTieSurvivor &&
      tieMerged == tieSecond->getHigh() && tieMerged->numInstances() == 2 &&
      tieMerged->getInstance(0) == tieFirst && tieMerged->getInstance(1) == tieSecond;
  bool marksClear = true;
  set<HighVariable *> observed;
  for(auto iter=fd.beginLoc();iter!=fd.endLoc();++iter) {
    HighVariable *high = (*iter)->getHigh();
    if (observed.insert(high).second)
      marksClear &= !high->isMark();
  }
  cout << "scale=" << freeCount
       << ",free_highs=" << freeHighs.size()
       << ",free_singletons=" << freeSingletons
       << ",legal_merged=" << legalMerged
       << ",sorted_survivor_unique=" << survivor
       << ",legal_instances=" << merged->numInstances()
       << ",instance_spaces=" << instanceSpaces(merged)
       << ",merge_groups=" << uniqueValue->getMergeGroup() << '/'
       << ramValue->getMergeGroup() << '/' << registerValue->getMergeGroup()
       << ",filters_singleton=" << filtersSingleton
       << ",type_identity_separate=" << identitySeparate
       << ",same_address_stable=" << sameAddressStable
       << ",marks_clear=" << marksClear << '\n';
}

} // namespace

int main()
{
  vector<string> specPaths;
  startDecompilerLibrary(specPaths);
  try {
    runScale(1);
    runScale(2);
    runScale(32);
    runScale(1024);
  }
  catch(const LowlevelError &error) {
    cerr << error.explain << '\n';
    shutdownDecompilerLibrary();
    return 1;
  }
  shutdownDecompilerLibrary();
  return 0;
}

File: tests/oracle/merge_datatype_scale_1204.rs
//! MERGE-DATATYPE-SCALE-0001 Rugra comparand for full-loc MergeType.

use rugra::address::Address;
use rugra::block::{BlockBasic, FlowBlock};
use rugra::cover::Cover;
use rugra::funcdata::Funcdata;
use rugra::merge::Merge;
use rugra::op::PcodeOpRef;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeMetatype};
use rugra::variable::HighVariable;
use rugra::varnode::{varnode_flags, Varnode};
use std::collections::HashSet;
use std::sync::{Arc, RwLock};

fn point_cover(value: &Arc<RwLock<Varnode>>, operation: &PcodeOpRef) {
    let order = operation.0.read().unwrap().get_seq_num().order;
    let mut cover = Cover::new();
    cover.add_def_point(0, order);
    cover.add_ref_point(0, order);
    let mut value = value.write().unwrap();
    value.cover = Some(Box::new(cover));
    value.clear_flags(varnode_flags::COVERDIRTY);
}

fn make_written(
    fd: &mut Funcdata,
    block: &Arc<RwLock<dyn FlowBlock + Send + Sync>>,
    space: AddressSpace,
    offset: u64,
    pc: u64,
    datatype: Arc<Datatype>,
) -> Arc<RwLock<Varnode>> {
    let operation = fd.new_op(0, Address::new(pc));
    fd.op_set_opcode(&operation, OpCode::CPUI_COPY);
    let value = fd
        .vbank
        .create_def_with_space(4, space, offset, &operation.0);
    value.write().unwrap().update_type(datatype);
    operation.0.write().unwrap().output = Some(value.clone());
    fd.op_insert_end(&operation, block);
    point_cover(&value, &operation);
    value
}

fn instance_spaces(high: &Arc<RwLock<HighVariable>>) -> String {
    high.read()
        .unwrap()
        .instances
        .iter()
        .map(|value| value.read().unwrap().address_space.space_id().to_string())
        .collect::<Vec<_>>()
        .join("/")
}

fn run_scale(free_count: usize) {
    let mut fd = Funcdata::new("merge_scale", Address::new(0x8000), 0x100);
    let block: Arc<RwLock<dyn FlowBlock + Send + Sync>> =
        Arc::new(RwLock::new(BlockBasic::new(0, Address::new(0x8000))));
    fd.bblocks.add_block(block.clone());

    let common = Arc::new(Datatype::Base(TypeBase::new(
        "uint4".to_string(),
        4,
        TypeMetatype::Uint,
    )));
    let identity_a = Arc::new(Datatype::Base(TypeBase::new(
        "identity_same".to_string(),
        4,
        TypeMetatype::Uint,
    )));
    let identity_b = Arc::new(Datatype::Base(TypeBase::new(
        "identity_same".to_string(),
        4,
        TypeMetatype::Uint,
    )));
    let tie_type = Arc::new(Datatype::Base(TypeBase::new(
        "same_address_tie".to_string(),
        4,
        TypeMetatype::Uint,
    )));

    let mut free_values = Vec::new();
    for index in 0..free_count {
        let value = fd.vbank.create_with_space(
            4,
            AddressSpace::Register,
            0x1000 + index as u64 * 8,
        );
        value.write().unwrap().update_type(common.clone());
        free_values.push(value);
    }

    let unique_value = make_written(
        &mut fd,
        &block,
        AddressSpace::Unique,
        0x30,
        0x8000,
        common.clone(),
    );
    let ram_value = make_written(
        &mut fd,
        &block,
        AddressSpace::Ram,
        0x20,
        0x8001,
        common.clone(),
    );
    let register_value = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x10,
        0x8002,
        common.clone(),
    );
    let implied = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x40,
        0x8003,
        common.clone(),
    );
    implied.write().unwrap().set_implied();
    let proto_partial = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x50,
        0x8004,
        common.clone(),
    );
    proto_partial.write().unwrap().set_proto_partial();
    let spacebase = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x60,
        0x8005,
        common,
    );
    spacebase
        .write()
        .unwrap()
        .set_flags(varnode_flags::SPACEBASE);
    let type_a = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x70,
        0x8006,
        identity_a,
    );
    let type_b = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x80,
        0x8007,
        identity_b,
    );
    let tie_first = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x90,
        0x8008,
        tie_type.clone(),
    );
    let tie_second = make_written(
        &mut fd,
        &block,
        AddressSpace::Register,
        0x90,
        0x8009,
        tie_type,
    );

    fd.set_high_level();
    let expected_survivor = unique_value.read().unwrap().high.clone().unwrap();
    let expected_tie_survivor = tie_first.read().unwrap().high.clone().unwrap();
    Merge::new().merge_by_datatype(&mut fd);

    let free_highs: HashSet<usize> = free_values
        .iter()
        .map(|value| {
            let high = value.read().unwrap().high.clone().unwrap();
            Arc::as_ptr(&high) as usize
        })
        .collect();
    let free_singletons = free_values.iter().all(|value| {
        value
            .read()
            .unwrap()
            .high
            .as_ref()
            .unwrap()
            .read()
            .unwrap()
            .instances
            .len()
            == 1
    });
    let merged = unique_value.read().unwrap().high.clone().unwrap();
    let legal_merged = [ram_value.clone(), register_value.clone()].iter().all(|value| {
        let high = value.read().unwrap().high.clone().unwrap();
        Arc::ptr_eq(&merged, &high)
    });
    let filters_singleton = [implied, proto_partial, spacebase].iter().all(|value| {
        value
            .read()
            .unwrap()
            .high
            .as_ref()
            .unwrap()
            .read()
            .unwrap()
            .instances
            .len()
            == 1
    });
    let type_identity_separate = {
        let a = type_a.read().unwrap().high.clone().unwrap();
        let b = type_b.read().unwrap().high.clone().unwrap();
        !Arc::ptr_eq(&a, &b)
    };
    let same_address_stable = {
        let high = tie_first.read().unwrap().high.clone().unwrap();
        let second_high = tie_second.read().unwrap().high.clone().unwrap();
        let instances = high.read().unwrap().instances.clone();
        Arc::ptr_eq(&high, &expected_tie_survivor)
            && Arc::ptr_eq(&high, &second_high)
            && instances.len() == 2
            && Arc::ptr_eq(&instances[0], &tie_first)
            && Arc::ptr_eq(&instances[1], &tie_second)
    };
    let mut observed = HashSet::new();
    let marks_clear = fd.vbank.loc_tree.iter().all(|entry| {
        let high = entry.0.read().unwrap().high.clone().unwrap();
        let key = Arc::as_ptr(&high) as usize;
        !observed.insert(key) || !high.read().unwrap().is_mark()
    });
    let groups = (
        unique_value.read().unwrap().mergegroup,
        ram_value.read().unwrap().mergegroup,
        register_value.read().unwrap().mergegroup,
    );
    println!(
        "scale={free_count},free_highs={},free_singletons={},legal_merged={},sorted_survivor_unique={},legal_instances={},instance_spaces={},merge_groups={}/{}/{},filters_singleton={},type_identity_separate={},same_address_stable={},marks_clear={}",
        free_highs.len(),
        u8::from(free_singletons),
        u8::from(legal_merged),
        u8::from(Arc::ptr_eq(&merged, &expected_survivor)),
        merged.read().unwrap().instances.len(),
        instance_spaces(&merged),
        groups.0,
        groups.1,
        groups.2,
        u8::from(filters_singleton),
        u8::from(type_identity_separate),
        u8::from(same_address_stable),
        u8::from(marks_clear),
    );
}

fn main() {
    for free_count in [1usize, 2, 32, 1024] {
        run_scale(free_count);
    }
}
